iceoryx_dust/test/moduletests/test_vocabulary_span.cpp

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) 2022 - 2023 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iox/span.hpp"
+#include "iox/vector.hpp"
+#include "test.hpp"
+
+#include <vector>
+
+namespace
+{
+using namespace ::testing;
+using namespace iox;
+
+TEST(span_test, NewEmptySpanCreatedFromIteratorContainsSameData)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "59980664-c94f-4bb5-bc9d-adeac630746e");
+    constexpr int* kNull = nullptr;
+
+    constexpr span<int> empty_sut(kNull, 0);
+
+    ASSERT_TRUE(empty_sut.empty());
+    EXPECT_EQ(nullptr, empty_sut.data());
+}
+
+TEST(span_test, NewDynSpanCreatedFromIteratorContainsSameData)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "05db30c2-e13d-4116-ba05-668b30ba4a23");
+    std::vector<int> vector = {1, 1, 2, 3, 5, 8};
+
+    span<int> dyn_sut(vector.begin(), vector.size());
+
+    EXPECT_EQ(vector.data(), dyn_sut.data());
+    EXPECT_EQ(vector.size(), dyn_sut.size());
+    for (size_t i = 0; i < dyn_sut.size(); ++i)
+    {
+        EXPECT_EQ(vector[i], dyn_sut[i]);
```

Comment:
While I like the Fibonacci sequence it would be good to have an initial check whether the creation of the `span` does not corrupt the container. With the current check this is not ensured. It would be nice to have e.g. a vector from 0 to 5 and to loop could also check for 0 to 5. Additionally, the data pointer and size of the vector could be obtained before the `span` is created. Maybe I'm a bit paranoid but one does never know :sweat_smile: ... and it has to be done only once ... alternatively, there could be a
```cpp
const std::vector<int> expectedData = {1, 1, 2, 3, 5, 8};
std::vector<int> vector = expectedData;
```
and the checks could be performed with `expectedData`

Reply:
Done.